### ISSUE TITLE
OrderBy extension method

### ISSUE THREAD

**Issue by callanh** [sq/JSIL#878]
What's the state of play for the OrderBy extension method? I have just run into this error:

The external method 'System.Linq.IOrderedEnumerable`1[!!0] OrderBy<TSource, TKey> (System.Collections.Generic.IEnumerable`1[!!0], System.Func`2[!!0, !!1])' of type 'System.Linq.Enumerable' has not been implemented.

Some advice would be appreciated.

**Comment by iskiselev** [comment]
Looks like it was not implemented in JS libraries. You have two options here: implement it yourself or use version translated from original .net BCL assembly.
I worked a lot on using always translated version of BCL assemblies, and it works for me - but it is not yet officially supported configuration, so YMMV and besides there is still a little bit gray area with licensing.
Please look on discussion in #843, #592, #336 and feel free ask me questions :).
You can also look on tests in SimpleTest project, you will be interested in `SimpleTestCasesForTranslatedBcl`.

**Comment by kg** [comment]
For now my recommendation would be to try out the translated BCL. The odds of the entire set of LINQ APIs being implemented in the hand-written JS bootstrap are pretty low. But if necessary, you can definitely implement that method - some of the others are already there (https://github.com/sq/JSIL/blob/master/JSIL.Libraries/Includes/Bootstrap/Linq/Classes/System.Linq.Enumerable.js etc)

**Comment by callanh** [comment]
When you say 'implement it yourself' what do you mean?  I am happy with a pragmatic solution.  I'd rather have a suboptimal implementation of OrderBy than have to ban OrderBy from my coding patterns.

**Comment by iskiselev** [comment]
I mean that you could provide your own implementation of OrderBy in *.js files. To do it easy, you could implement OrderBy (with same signature you use in your code) within some small test app, translate it, take translated implementation and insert it in your (or library) *.js file as implementation of `Enumerable.OrderBy<>(...)`.

I personally could advice you to try another option and use translated BCL instead (see #843 for info), but first solution may also work based on size of your app and number of different BCL API used.

**Comment by iskiselev** [comment]
You can also look https://github.com/sq/JSIL/wiki/Troubleshooting-FAQ. There is an description how to provide your own implementation for library method.

**Comment by callanh** [comment]
Thank you for your advice and I hope you don't mind a few more questions.

This is the reflected .NET code:

```C#
    public static IOrderedEnumerable<TSource> OrderBy<TSource, TKey>(this IEnumerable<TSource> source, Func<TSource, TKey> keySelector)
    {
      return (IOrderedEnumerable<TSource>) new OrderedEnumerable<TSource, TKey>(source, keySelector, (IComparer<TKey>) null, false);
    }
````

This is my attempt at a JS version inside JSIL.Bootstrap.Linq.js

```js
$.Method({ Static: true, Public: true }, "OrderBy",
      new JSIL.MethodSignature($jsilcore.TypeRef("System.Collections.Generic.IEnumerable`1", ["!!0"]), [$jsilcore.TypeRef("System.Collections.Generic.IEnumerable`1", ["!!0"]), $jsilcore.TypeRef("System.Func`2", ["!!0", "!!1"])], ["TSource", "TKey"]),
      function OrderBy$b2(TSource, TKey, source, keySelector) {
          return new (System.Linq.OrderedEnumerable$b2.Of(TSource, TKey))(source, keySelector, null, false);
      }
      );
```

Am I even close?  It still fails with the same message so I'm definitely missing something.

**Comment by iskiselev** [comment]
Looks like method should return `'System.Linq.IOrderedEnumerable<T>` based on your warning. In some place it may be critical to use same signature.

**Comment by callanh** [comment]
Good spot, second attempt still fails saying it's not implemented.  Are there any other places to update?

```JS
      $.Method({ Static: true, Public: true }, "OrderBy",
      new JSIL.MethodSignature($jsilcore.TypeRef("System.Linq.IOrderedEnumerable`1", ["!!0"]), [$jsilcore.TypeRef("System.Collections.Generic.IEnumerable`1", ["!!0"]), $jsilcore.TypeRef("System.Func`2", ["!!0", "!!1"])], ["TSource", "TKey"]),
      function OrderBy$b2(TSource, TKey, source, keySelector) {
        return new (System.Linq.OrderedEnumerable$b2.Of(TSource, TKey))(source, keySelector, null, false);
      }
      );
```

**Comment by iskiselev** [comment]
Right now I don't see problems. I will look on it tomorrow if you will not report success.
My final advises for today:
- Open translated mscorlib *.js file, find signature of this method (it will be defined with $.ExternalMethod in Enumerable) and copy it.
- Create test with OrderBy in SimpleTests\ folder - it will be much easy for you to test. Don't forget to build solution after each change in JSIL.Libraries project (test run will not force it). You always can debug translated test in browser - if test fail, output will contain a link to open for debug.

**Comment by iskiselev** [comment]
@callanh, sorry, haven't looked on it today. Will try tomorrow.

**Comment by callanh** [comment]
@iskiselev thanks for letting me know (and offering to help in the first place).

**Comment by iskiselev** [comment]
I get working version of signature in SimpleTest. For it I edited JSIL.Libraries\Includes\Linq\Classes\System.Linq.Enumerable.js. Here is patch:
```
@@ -530,6 +530,13 @@
             return source;
         }
       );
+
+    $.Method({ Static: true, Public: true }, "OrderBy",
+      new JSIL.MethodSignature($jsilcore.TypeRef("System.Linq.IOrderedEnumerable`1", ["!!0"]), [$jsilcore.TypeRef("System.Collections.Generic.IEnumerable`1", ["!!0"]), $jsilcore.TypeRef("System.Func`2", ["!!0", "!!1"])], ["TSource", "TKey"]),
+      function OrderBy$b2(TSource, TKey, source, keySelector) {
+        throw new Error("AAA");
+      }
+    );
   }
 );
 
@@ -557,4 +564,9 @@
     $.ExternalMethod({ Static: true, Public: true }, "ToArray",
       new JSIL.MethodSignature($jsilcore.TypeRef("System.Array", ["!!0"]), [$jsilcore.TypeRef("System.Collections.Generic.IEnumerable`1", ["!!0"])], ["TSource"])
     );
+
+    $.ExternalMethod({ Static: true, Public: true }, "OrderBy",
+      new JSIL.MethodSignature($jsilcore.TypeRef("System.Linq.IOrderedEnumerable`1", ["!!0"]), [$jsilcore.TypeRef("System.Collections.Generic.IEnumerable`1", ["!!0"]), $jsilcore.TypeRef("System.Func`2", ["!!0", "!!1"])], ["TSource", "TKey"])
+    );
+
 });
@@ -560,1 +572,5 @@
 });
+
+JSIL.MakeInterface(
+  "System.Linq.IOrderedEnumerable`1", true, ["TKey"], function ($) {
+  }, []);
\ No newline at end of file
```

**Comment by callanh** [comment]
Thanks @iskiselev, now to work out how to implement this method!

**Comment by callanh** [comment]
It seemed like System.Linq.OrderedEnumerable class was already implemented in JSIL?  I was thinking I could just implement as reflected from the .NET code:

```C#
return (IOrderedEnumerable<TSource>) new OrderedEnumerable<TSource, TKey>(source, keySelector, (IComparer<TKey>) null, false);
```

I've attempted to convert to JS but it fails with: JavaScript runtime error: Unable to get property 'Of' of undefined or null reference.

```JS
return new (System.Linq.OrderedEnumerable$b2.Of(TSource, TKey))(source, keySelector, null, false);
```

I'm not sure if it is my assumptions or my JS skills that are flawed in this case.

**Comment by kg** [comment]
It would be ```$b1``` because it has one generic parameter, not ```$b2```

**Comment by callanh** [comment]
It's meant to be creating a new:

```
OrderedEnumerable<TSource, TKey>
```

and casting that in the return as:

```
(IOrderedEnumerable<TSource>)
```

It's seems the JS conversion has dropped the typecast because it's implicit anyway?

**Comment by kg** [comment]
Oh, I misread. If it says that type isn't defined, then it's not defined. :(

**Comment by callanh** [comment]
Right, I was confused by the existence of the header code in the generated System.Core.js.

/* class System.Linq.OrderedEnumerable`2 */ 

So that means this class and all its dependencies need to be manually implemented in System.Linq.Enumerable.js?

**Comment by kg** [comment]
Yeah. I think translated BCL is the right solution for you here, TBH.

**Comment by callanh** [comment]
I can confirm that using the translated BCL gives a working implementation of ```System.Linq.Enumerable.OrderBy<TSource, TKey>```

**Comment by iskiselev** [comment]
Look #332,  in general you should add to your *.jsilconfig something like:
```
  "AnalyzerSettings": {
    "DeadCodeAnalyzer": {
      "DeadCodeElimination": true,
      "WhiteList": [
        "System\\.Void Foo\\.SomeClass::\\WhiteListedMethod()",
        "Bar\\.Whitelisted\\.Class",
      ]
    }
  }
```
All parameters and meta attributes that manage have some description in wiki: https://github.com/sq/JSIL/wiki/Configuration-Files

**Comment by kg** [comment]
You shouldn't need to whitelist too many things; let us know if you find that you have to whitelist a bunch of stuff - might indicate a missing feature.

**Comment by callanh** [comment]
After turning on dead code elimination my app is having this issue:

```
The external method 'System.Boolean UseRandomizedHashing()' of type 'System.String' has not been implemented.
```

I'm inside a static class constructor, using reflection to set the properties of the static class, from data loaded from an embedded resource.  Basically, it's a PCL implementation of Resources.

Any thoughts or suggestions?

**Comment by kg** [comment]
I don't think String is translated, so it probably translated something that uses that method. Can you figure out what's calling it?

**Comment by callanh** [comment]
This seems to be the only call to UseRandomizedHashing.  It's not clear what is calling onto HashHelpers though.

```JS
function HashHelpers__cctor () {
      $thisType.s_UseRandomizedStringHashing = $T00().UseRandomizedHashing();
      $thisType.primes = JSIL.Array.New($T01(), [3, 7, 11, 17, 23, 29, 37, 47, 59, 71, 89, 107, 131, 163, 197, 239, 293, 353, 431, 521, 631, 761, 919, 1103, 1327, 1597, 1931, 2333, 2801, 3371, 4049, 4861, 5839, 7013, 8419, 10103, 12143, 14591, 17519, 21023, 25229, 30293, 36353, 43627, 52361, 62851, 75431, 90523, 108631, 130363, 156437, 187751, 225307, 270371, 324449, 389357, 467237, 560689, 672827, 807403, 968897, 1162687, 1395263, 1674319, 2009191, 2411033, 2893249, 3471899, 4166287, 4999559, 5999471, 7199369]);
      $thisType.currentIndex = 1024;
      $thisType.lockObj = new ($T02())();
    }; 
```

**Comment by kg** [comment]
It's probably sufficient to just stub out UseRandomizedHashing to return false. It should be easy to add, just find the System.String external methods. There's a handful of them.

**Comment by callanh** [comment]
Would adding it to the whitelist be better?

**Comment by kg** [comment]
No, because I don't think System.String gets translated

**Comment by iskiselev** [comment]
Use something like: https://github.com/iskiselev/JSIL/commit/309d5efb79761855cf6c10d0c6b292f0d36d8444

**Comment by callanh** [comment]
Thanks for the patch, that gets me past that issue.

The next one is due to my serialization code

```
System.Exception: Type does not have a default constructor
```

So the problem is that .NET requires a default constructor if you want to create an instance with reflection.

So my persisted classes all have default public constructors that are empty, just so that I can create them with reflection.

I think DCE is stripping them out - is there a way to keep these empty constructors?

**Comment by kg** [comment]
Hmmm... DCE probably shouldn't strip empty constructors. Not sure how to whitelist them off the top of my head.

**Comment by callanh** [comment]
There doesn't seem to be a setting flag... at least according to the wiki documentation.

**Comment by iskiselev** [comment]
DCE will strip empty any ctro that is not used statically. To use them through reflection, you should white-list them either by meta - attribute or by regexp in jsilconfig.

**Comment by callanh** [comment]
Hmm, not really viable for my use case - can you point me to the DCE code that makes this decision?

**Comment by iskiselev** [comment]
@callanh, add me on skype: igor_kiselev and I'll try to help you.

**Comment by kg** [comment]
I think we probably want an option to just globally ensure that any non-pruned type keeps its default ctor (if it has one). Doesn't need to be the default.

**Comment by iskiselev** [comment]
You could white-list all your classes by namespace .

**Comment by iskiselev** [comment]
@kg, not sure about option. You need "default" ctor only when you start serialize/deserialize, but in that case you will probably want preserve all fields also.
Another trick that could be done - add method that will use all "default" ctors (that should be saved) and mark this method as preserved.

**Comment by callanh** [comment]
I'd be happy with an option to keep the default ctors.  I can't comment on whether this should be on/off by default.

**Comment by kg** [comment]
Igor's suggestion of a single method that uses all the default ctors seems the most manageable. Do you mind trying that out?

**Comment by callanh** [comment]
Igor convinced me to use white list by namespace by informing me that DCE will eliminate unused fields.  I have a lot of classes for serialization and they can't be reduced this way.  The good news is this is all working well now.

So I've switched to translated BCL with DCE to reduce the overall compile time to around ~2 minutes.  Compared to about 15 seconds using the stubbed config.

Two minutes is okay for production builds, but I'd like to get back down below 15 seconds for development builds if there is anything more I can try?

**Comment by kg** [comment]
There aren't any obvious ways to make it faster. If you do a CPU profiler of the compiler while it's at work on your application, maybe something there will stand out that I can fix... but historically something like 50% of the cpu time is spent in ILSpy and 20-30% is in the garbage collector. Not much I can fix easily.

**Comment by callanh** [comment]
Well, building the JSIL Compiler for RELEASE instead of DEBUG was a good start... oops.  Down to 53 seconds vs 11 seconds.
